web/modules/custom/joinup_core/src/WorkflowHelper.php

```diff
@@ -5,14 +5,90 @@
 use Drupal\Component\Plugin\PluginInspectionInterface;
 use Drupal\Core\Entity\FieldableEntityInterface;
 use Drupal\Core\Field\FieldDefinitionInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\Session\AccountProxyInterface;
+use Drupal\Core\Session\AccountSwitcherInterface;
 use Drupal\state_machine\Plugin\Workflow\WorkflowInterface;
+use Drupal\state_machine\Plugin\Workflow\WorkflowTransition;
 
 /**
  * Contains helper methods to retrieve workflow related data from entities.
  */
 class WorkflowHelper implements WorkflowHelperInterface {
 
   /**
+   * The account switcher service.
+   *
+   * @var \Drupal\Core\Session\AccountSwitcherInterface
+   */
+  protected $accountSwitcher;
+
+  /**
+   * The current user proxy.
+   *
+   * @var \Drupal\Core\Session\AccountProxyInterface
+   */
+  protected $currentUser;
+
+  /**
+   * Constructs a WorkflowHelper.
+   *
+   * @param \Drupal\Core\Session\AccountProxyInterface $currentUser
+   *   The service that contains the current user.
+   * @param \Drupal\Core\Session\AccountSwitcherInterface $accountSwitcher
+   *   The account switcher interface.
+   */
+  public function __construct(AccountProxyInterface $currentUser, AccountSwitcherInterface $accountSwitcher) {
+    $this->accountSwitcher = $accountSwitcher;
+    $this->currentUser = $currentUser;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAvailableStatesLabels(FieldableEntityInterface $entity, AccountInterface $user = NULL) {
+    $allowed_transitions = $this->getAvailableTransitions($entity, $user);
+
+    $allowed_states = array_map(function (WorkflowTransition $transition) {
+      return (string) $transition->getToState()->getLabel();
+    }, $allowed_transitions);
+
+    return $allowed_states;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAvailableTransitionsLabels(FieldableEntityInterface $entity, AccountInterface $user = NULL) {
+    return array_map(function (WorkflowTransition $transition) {
+      return (string) $transition->getLabel();
+    }, $this->getAvailableTransitions($entity, $user));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAvailableTransitions(FieldableEntityInterface $entity, AccountInterface $user = NULL) {
+    // Set the current user so that states available are retrieved for the
+    // specific account.
+    $account_switched = FALSE;
+    if ($user !== NULL && $user->id() !== $this->currentUser->id()) {
+      $this->accountSwitcher->switchTo($user);
+      $account_switched = TRUE;
+    }
```

Comment:
The proper solution would be to be able to pass the `$account` to the State Machine when retrieving the transitions. This switching of the current user is rather tricky. Let's put a `@todo` here that explains this is a workaround, with the instructions that it should be removed once the upstream issue is fixed: https://www.drupal.org/node/2776969